cpp/src/arrow/io/interfaces.cc

```diff
@@ -26,18 +26,27 @@
 namespace arrow {
 namespace io {
 
-FileInterface::~FileInterface() {}
+FileInterface::~FileInterface() = default;
 
-RandomAccessFile::RandomAccessFile() { set_mode(FileMode::READ); }
+struct RandomAccessFile::RandomAccessFileImpl {
+  std::mutex lock_;
+};
+
+RandomAccessFile::~RandomAccessFile() = default;
+
+RandomAccessFile::RandomAccessFile() :
+  impl_(new RandomAccessFile::RandomAccessFileImpl()) {}
 
 Status RandomAccessFile::ReadAt(int64_t position, int64_t nbytes, int64_t* bytes_read,
                                 uint8_t* out) {
+  std::lock_guard<std::mutex> lock(impl_->lock_);
   RETURN_NOT_OK(Seek(position));
   return Read(nbytes, bytes_read, out);
 }
 
 Status RandomAccessFile::ReadAt(int64_t position, int64_t nbytes,
                                 std::shared_ptr<Buffer>* out) {
+  std::lock_guard<std::mutex> lock(impl_->lock_);
```

Comment:
If we're going to do this, then we can remove a couple of the other ReadAt implementations in subclasses which are basically copies of this

Reply:
Oh so one nuance here we need to be careful of is that some of the subclasses have their own mutexes. So we need to be using the same mutex in all methods that may cause race conditions, so we may not want to make any changes

Reply:
To remedy this I made the `ReadAt` method pure virtual, so `RandomAccessFile` provides thread safe implementations of these methods which must be explicitly overridden by child classes. Those child classes can directly call the `RandomAccessFile` implementations so we get thread safety by default and folks have to explicitly use the base classes methods.

Reply:
I'm pretty surprised that you can make the methods pure virtual but also have a default implementation, I guess it works?

Reply:
Documentation for it is here (kind of easy to miss): http://en.cppreference.com/w/cpp/language/abstract_class:

> The definition of a pure virtual function may be provided (and must be provided if the pure virtual is the destructor):